internal/kernel_SSE.h

```diff
@@ -0,0 +1,332 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// kernel_SSE.h: a collection of Intel SSE optimized kernels.
+// Check in kernel_default.h which one(s) are actually used by default.
+// Others are mere experiments; they are still covered by tests
+// in case they might be useful some day.
+// Writen by: Sagi Marovich, mail: sagi.marcovich@intel.com
+
+#ifndef GEMMLOWP_INTERNAL_KERNEL_SSE_H_
+#define GEMMLOWP_INTERNAL_KERNEL_SSE_H_
+
+
+#include "kernel.h"
+#include <cassert>
+#include <string.h>
+#include <iostream>
+
+
+
+
+namespace gemmlowp {
+
+#ifdef GEMMLOWP_SSE32
+struct SSE32Kernel4x4Depth2 : KernelBase {
+  typedef KernelFormat<KernelSideFormat<CellFormat<4, 2,CellOrder::WidthMajor>, 1>,
+                       KernelSideFormat<CellFormat<4, 2,CellOrder::WidthMajor>, 1> > Format;					   
+
+
+  const char* Name() const override { return "SSE, 4x4, depth 2"; }
+
+  void Run(std::int32_t* dst_ptr, int dst_row_stride, int dst_col_stride,
+           const std::uint8_t* lhs_ptr, const std::uint8_t* rhs_ptr,
+           int start_depth, int run_depth) const override {
+		   
+
+	
+	ScopedProfilingLabel label("optimized kernel");
+	assert(dst_row_stride == 1); 
+	std::int32_t run_depth_cells = run_depth / Format::kDepth;
+	
+	
+	/* Main loop */
+
+
+	// A 2x4 cell of Rhs is stored in 16bit in xmm1 .
+	// A 4x2 block Lhs is stored in 16bit in xmm0.
+	// A 4x4 block of accumulators is stored in 32bit in xmm4--xmm7.
+	//
+	//                   +-------+-------+-------+-------+
+	//                   |xmm1[0]|xmm1[2]|xmm1[4]|xmm1[6]|
+	//              Rhs  +-------+---------------+-------+
+	//                   |xmm1[1]|xmm1[3]|xmm1[5]|xmm1[7]|
+	//                   +-------+-------+-------+-------+
+	//
+	//                   |       |       |       |       |
+	//                                                   
+	//    Lhs            |       |       |       |       |
+	//
+	//  +--+--+ - - - -  +-------+-------+-------+-------+
+	//  |xmm0 |          | xmm4  | xmm5  | xmm6  | xmm7  |
+	//  |xmm0 | (Iter1)  | xmm4  | xmm5  | xmm6  | xmm7  |
+	//  |xmm0 |          | xmm4  | xmm5  | xmm6  | xmm7  |
+	//  |xmm0 |          | xmm4  | xmm5  | xmm6  | xmm7  |
+	//  +--+--+ - - - -  +-------+-------+-------+-------+
+	//
+	//                              Accumulator
+
+	
+	asm volatile(
+		
+		
+		//set accumulators to zero.
+		"pxor %%xmm4  , %%xmm4 \n\t" 
+		"pxor %%xmm5  , %%xmm5 \n\t"
+		"pxor %%xmm6  , %%xmm6 \n\t" 
+		"pxor %%xmm7  , %%xmm7 \n\t"	
+		
+		"outerLoop%=:\n\t"
+		
+		//RHS cell to xmm1
+		"pmovzxbw (%[rhs_ptr]), %%xmm1\n\t"
+		
+		//LHS cell
+		"pmovzxbw 0x00(%[lhs_ptr]), %%xmm0\n\t"	
+		"pshufd $0x00,%%xmm1,%%xmm2     \n\t"
+		"pshufd $0x55,%%xmm1,%%xmm3     \n\t"
+		"pmaddwd %%xmm0, %%xmm2         \n\t"
+		"pmaddwd %%xmm0, %%xmm3         \n\t"
+		"paddd %%xmm2, %%xmm4           \n\t"
+		"paddd %%xmm3, %%xmm5           \n\t"
```

Comment:
Thanks. I look forward to future improvements on these kernels and am already very grateful for these as-is :-)